Failing to build opm-common (master) on macOS (gcc 9 or Apple Clang 10)

I get a compiler error when trying to build opm-common:

```
CMakeFiles/genkw.dir/src/opm/parser/eclipse/Parser/createDefaultKeywordList.cpp.o
In file included from /usr/local/Cellar/gcc/9.2.0_1/include/c++/9.2.0/system_error:39,
                 from /usr/local/Cellar/gcc/9.2.0_1/include/c++/9.2.0/bits/ios_base.h:46,
                 from /usr/local/Cellar/gcc/9.2.0_1/include/c++/9.2.0/ios:42,
                 from /usr/local/Cellar/gcc/9.2.0_1/include/c++/9.2.0/ostream:38,
                 from /usr/local/Cellar/gcc/9.2.0_1/include/c++/9.2.0/iostream:39,
                 from /Users/pumbaa/dumux-course/dumux/opm-common/src/opm/parser/eclipse/Parser/createDefaultKeywordList.cpp:25:
/usr/local/Cellar/gcc/9.2.0_1/include/c++/9.2.0/x86_64-apple-darwin17/bits/error_constants.h:135:24: error: 'EOWNERDEAD' was not declared in this scope
  135 |       owner_dead =     EOWNERDEAD,
      |                        ^~~~~~~~~~
/usr/local/Cellar/gcc/9.2.0_1/include/c++/9.2.0/x86_64-apple-darwin17/bits/error_constants.h:151:34: error: 'ENOTRECOVERABLE' was not declared in this scope
  151 |       state_not_recoverable =    ENOTRECOVERABLE,
      |                                  ^~~~~~~~~~~~~~~
```

below the entire cmake log until the error (which also includes my system specifics):

```
cmake   -DCMAKE_C_COMPILER=gcc-9 -DCMAKE_CXX_COMPILER=g++-9 -DCMAKE_PREFIX_PATH='/usr/local/opt/zlib' -DCMAKE_CXX_FLAGS_RELEASE=' -fno-strict-aliasing -fstrict-overflow -fno-finite-math-only -O3 -march=native -funroll-loops -g0 -Wall -Wno-deprecated -Wno-deprecated-declarations -Wno-sign-compare' -DCMAKE_CXX_FLAGS_DEBUG='-O0 -ggdb -Wall' -DCMAKE_BUILD_TYPE=Release -DDUNE_PYTHON_VIRTUALENV_SETUP=1 -DDUNE_PYTHON_FORCE_PYTHON3=1 "-Ddune-common_DIR=/Users/pumbaa/dumux-course/dumux/dune-common/build-cmake" -DCMAKE_C_COMPILER=gcc-9 -DCMAKE_CXX_COMPILER=g++-9 -DCMAKE_PREFIX_PATH='/usr/local/opt/zlib' -DCMAKE_CXX_FLAGS_RELEASE=' -fno-strict-aliasing -fstrict-overflow -fno-finite-math-only -O3 -march=native -funroll-loops -g0 -Wall -Wno-deprecated -Wno-deprecated-declarations -Wno-sign-compare' -DCMAKE_CXX_FLAGS_DEBUG='-O0 -ggdb -Wall' -DCMAKE_BUILD_TYPE=Release -DDUNE_PYTHON_VIRTUALENV_SETUP=1 -DDUNE_PYTHON_FORCE_PYTHON3=1 "/Users/pumbaa/dumux-course/dumux/opm-common"
CMake Warning at cmake/Modules/Findopm-tests.cmake:26 (message):
  opm-tests not found - integration tests using opm-tests will be skipped.
Call Stack (most recent call first):
  CMakeLists.txt:59 (include)


-- Found Boost: /usr/local/include (found suitable version "1.55.0", minimum required is "1.44.0") found components:  system unit_test_framework 
CMake Deprecation Warning at cmake/Modules/OpmLibMain.cmake:25 (cmake_policy):
  The OLD behavior for policy CMP0026 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.
Call Stack (most recent call first):
  CMakeLists.txt:156 (include)


CMake Deprecation Warning at cmake/Modules/OpmLibMain.cmake:29 (cmake_policy):
  The OLD behavior for policy CMP0048 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.
Call Stack (most recent call first):
  CMakeLists.txt:156 (include)


-- Performing Test HAVE_ATTRIBUTE_ALWAYS_INLINE
-- Performing Test HAVE_ATTRIBUTE_ALWAYS_INLINE - Success
-- Performing Test HAS_ATTRIBUTE_UNUSED
-- Performing Test HAS_ATTRIBUTE_UNUSED - Success
-- Performing Test HAS_ATTRIBUTE_DEPRECATED
-- Performing Test HAS_ATTRIBUTE_DEPRECATED - Success
-- Performing Test HAS_ATTRIBUTE_DEPRECATED_MSG
-- Performing Test HAS_ATTRIBUTE_DEPRECATED_MSG - Success
-- CMake version: 3.15.5
-- Mac OS X version: 10.13.6 (17G11023)
-- Target architecture: x86_64
-- Found Git: /usr/local/bin/git (found version "2.21.0") 
-- Source code repository: git d3817e2fc
-- GNU C++ compiler version: 9.2.0
-- Linker: @(#)PROGRAM:ld  PROJECT:ld64-409.12
BUILD 17:47:51 Sep 25 2018
configured to support archs: armv6 armv7 armv7s arm64 i386 x86_64 x86_64h armv6m armv7k armv7m armv7em
LTO support using: LLVM version 10.0.0, (clang-1000.10.44.4) (static support for 21, runtime is 21)
TAPI support using: Apple TAPI version 10.0.0 (tapi-1000.10.8)
-- Checking to see if CXX compiler accepts flag -Wl,--enable-new-dtags
-- Checking to see if CXX compiler accepts flag -Wl,--enable-new-dtags - no
-- Precompiled headers: disabled
-- Build type: Release
-- Checking to see if CXX compiler accepts flag -mtune=native
-- Checking to see if CXX compiler accepts flag -mtune=native - yes
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Checking to see if CXX compiler accepts flag -pthread
-- Checking to see if CXX compiler accepts flag -pthread - yes
-- Could NOT find CppCheck (missing: CPPCHECK_PROGRAM) 
-- Disabling clang-check as CMAKE_EXPORT_COMPILE_COMMANDS is not enabled
-- Looking for strip utility
-- Looking for strip utility - found
-- Performing Test HAVE_DYNAMIC_BOOST_TEST
-- Performing Test HAVE_DYNAMIC_BOOST_TEST - Failed
-- Writing config file "/Users/pumbaa/dumux-course/dumux/opm-common/build-cmake/config.h"...
CMake Warning (dev) at cmake/Modules/OpmInstall.cmake:43 (set):
  implicitly converting 'LOCATION' to 'STRING' type.
Call Stack (most recent call first):
  cmake/Modules/OpmLibMain.cmake:240 (opm_install)
  CMakeLists.txt:156 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- This build defaults to installing in /usr/local
-- Checking whether the file system is case-sensitive
-- File system is not case-sensitive
-- Found Doxygen: /usr/local/bin/doxygen (found version "1.8.17") found components:  doxygen dot 
-- Writing version information to local header project-version.h
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/pumbaa/dumux-course/dumux/opm-common/build-cmake
--- calling make for opm-common ---
build directory: build-cmake
cmake --build . -- 
Scanning dependencies of target genkw
[  1%] Building CXX object CMakeFiles/genkw.dir/src/opm/json/JsonObject.cpp.o
[  1%] Building CXX object CMakeFiles/genkw.dir/src/opm/parser/eclipse/Parser/createDefaultKeywordList.cpp.o
In file included from /usr/local/Cellar/gcc/9.2.0_1/include/c++/9.2.0/system_error:39,
                 from /usr/local/Cellar/gcc/9.2.0_1/include/c++/9.2.0/bits/ios_base.h:46,
                 from /usr/local/Cellar/gcc/9.2.0_1/include/c++/9.2.0/ios:42,
                 from /usr/local/Cellar/gcc/9.2.0_1/include/c++/9.2.0/ostream:38,
                 from /usr/local/Cellar/gcc/9.2.0_1/include/c++/9.2.0/iostream:39,
                 from /Users/pumbaa/dumux-course/dumux/opm-common/src/opm/parser/eclipse/Parser/createDefaultKeywordList.cpp:25:
/usr/local/Cellar/gcc/9.2.0_1/include/c++/9.2.0/x86_64-apple-darwin17/bits/error_constants.h:135:24: error: 'EOWNERDEAD' was not declared in this scope
  135 |       owner_dead =     EOWNERDEAD,
      |                        ^~~~~~~~~~
/usr/local/Cellar/gcc/9.2.0_1/include/c++/9.2.0/x86_64-apple-darwin17/bits/error_constants.h:151:34: error: 'ENOTRECOVERABLE' was not declared in this scope
  151 |       state_not_recoverable =    ENOTRECOVERABLE,
      |                                  ^~~~~~~~~~~~~~~
gmake[2]: *** [CMakeFiles/genkw.dir/build.make:76: CMakeFiles/genkw.dir/src/opm/parser/eclipse/Parser/createDefaultKeywordList.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:5303: CMakeFiles/genkw.dir/all] Error 2
```

AppleClang 10.0.0.10001044 gives me 

```
[  1%] Building CXX object CMakeFiles/genkw.dir/src/opm/json/JsonObject.cpp.o
In file included from /Users/pumbaa/dumux-course/dumux/opm-common/src/opm/json/JsonObject.cpp:26:
/Users/pumbaa/dumux-course/dumux/opm-common/opm/common/utility/FileSystem.hpp:24:10: fatal error: 'experimental/filesystem' file not found
#include <experimental/filesystem>
         ^~~~~~~~~~~~~~~~~~~~~~~~~
```

> ```
> /Users/pumbaa/dumux-course/dumux/opm-common/opm/common/utility/FileSystem.hpp:24:10: fatal error: 'experimental/filesystem' file not found
> #include <experimental/filesystem>
>          ^~~~~~~~~~~~~~~~~~~~~~~~~
> ```

That's weird.  You should only get [that line](https://github.com/OPM/opm-common/blob/d3817e2fc2f79014d3cb0c01325a6e8a5dcc0cf3/opm/common/utility/FileSystem.hpp#L24) if you're not compiling in C++17 mode.  Could you try rebuilding with the command
```sh
$ cmake --build . --target all -- -j1 VERBOSE=1
```
so that we may see the exact list of compiler options?

It contains `-std=c++1z` at least.

```
/usr/bin/clang++  -DHAVE_CONFIG_H=1 -I/Users/pumbaa/dumux-course/dumux/opm-common/build-cmake -I/Users/pumbaa/dumux-course/dumux/opm-common/external/cjson -I/Users/pumbaa/dumux-course/dumux/opm-common/build-cmake/include -I/Users/pumbaa/dumux-course/dumux/opm-common  -pipe -pthread -fno-strict-aliasing -fstrict-overflow -fno-finite-math-only -O3 -march=native -funroll-loops -g0 -Wall -Wno-deprecated -Wno-deprecated-declarations -Wno-sign-compare -mtune=native -DNDEBUG   -std=c++1z -o CMakeFiles/genkw.dir/src/opm/json/JsonObject.cpp.o -c /Users/pumbaa/dumux-course/dumux/opm-common/src/opm/json/JsonObject.cpp
In file included from /Users/pumbaa/dumux-course/dumux/opm-common/src/opm/json/JsonObject.cpp:26:
/Users/pumbaa/dumux-course/dumux/opm-common/opm/common/utility/FileSystem.hpp:24:10: fatal error: 'experimental/filesystem' file not found
#include <experimental/filesystem>
         ^~~~~~~~~~~~~~~~~~~~~~~~~
```

I goes on... A new clang 10 (this time not AppleClang 10 but llvm from homebrew) gives me

```
Scanning dependencies of target genkw
[  1%] Building CXX object CMakeFiles/genkw.dir/src/opm/json/JsonObject.cpp.o
In file included from /Users/pumbaa/dumux-course/dumux/opm-common/src/opm/json/JsonObject.cpp:26:
/Users/pumbaa/dumux-course/dumux/opm-common/opm/common/utility/FileSystem.hpp:36:47: warning: 'filesystem' is deprecated: std::experimental::filesystem has now been deprecated in favor of C++17's std::filesystem. Please stop using it and start using std::filesystem. This
      experimental version will be removed in LLVM 11. You can remove this warning by defining the _LIBCPP_NO_EXPERIMENTAL_DEPRECATION_WARNING_FILESYSTEM macro. [-Wdeprecated-declarations]
    namespace filesystem = std::experimental::filesystem;
                                              ^
/usr/local/opt/llvm/bin/../include/c++/v1/experimental/filesystem:246:1: note: 'filesystem' has been explicitly marked deprecated here
_LIBCPP_BEGIN_NAMESPACE_EXPERIMENTAL_FILESYSTEM
^
/usr/local/opt/llvm/bin/../include/c++/v1/experimental/__config:46:63: note: expanded from macro '_LIBCPP_BEGIN_NAMESPACE_EXPERIMENTAL_FILESYSTEM'
    _LIBCPP_BEGIN_NAMESPACE_EXPERIMENTAL namespace filesystem _LIBCPP_DEPRECATED_EXPERIMENTAL_FILESYSTEM { \
                                                              ^
/usr/local/opt/llvm/bin/../include/c++/v1/experimental/__config:42:70: note: expanded from macro '_LIBCPP_DEPRECATED_EXPERIMENTAL_FILESYSTEM'
#   define _LIBCPP_DEPRECATED_EXPERIMENTAL_FILESYSTEM __attribute__((deprecated("std::experimental::filesystem has now been deprecated in favor of C++17's std::filesystem. Please stop using it and start using std::filesystem. This experimental version will be remo...
                                                                     ^
In file included from /Users/pumbaa/dumux-course/dumux/opm-common/src/opm/json/JsonObject.cpp:28:
/Users/pumbaa/dumux-course/dumux/opm-common/opm/json/JsonObject.hpp:33:52: error: 'path' is unavailable: introduced in macOS 10.15
        explicit JsonObject(const Opm::filesystem::path& jsonFile );
                                                   ^
/usr/local/opt/llvm/bin/../include/c++/v1/filesystem:738:24: note: 'path' has been explicitly marked unavailable here
class _LIBCPP_TYPE_VIS path {
                       ^
/Users/pumbaa/dumux-course/dumux/opm-common/src/opm/json/JsonObject.cpp:51:51: error: 'path' is unavailable: introduced in macOS 10.15
    JsonObject::JsonObject(const Opm::filesystem::path& jsonFile ) {
                                                  ^
/usr/local/opt/llvm/bin/../include/c++/v1/filesystem:738:24: note: 'path' has been explicitly marked unavailable here
class _LIBCPP_TYPE_VIS path {
                       ^
/Users/pumbaa/dumux-course/dumux/opm-common/src/opm/json/JsonObject.cpp:52:39: error: 'string' is unavailable: introduced in macOS 10.15
        std::ifstream stream(jsonFile.string().c_str());
                                      ^
/usr/local/opt/llvm/bin/../include/c++/v1/filesystem:998:41: note: 'string' has been explicitly marked unavailable here
  _LIBCPP_INLINE_VISIBILITY std::string string() const { return __pn_; }
                                        ^
/Users/pumbaa/dumux-course/dumux/opm-common/src/opm/json/JsonObject.cpp:59:78: error: 'string' is unavailable: introduced in macOS 10.15
                throw std::invalid_argument("Parsing json file: " + jsonFile.string() + " failed ");
                                                                             ^
/usr/local/opt/llvm/bin/../include/c++/v1/filesystem:998:41: note: 'string' has been explicitly marked unavailable here
  _LIBCPP_INLINE_VISIBILITY std::string string() const { return __pn_; }
                                        ^
/Users/pumbaa/dumux-course/dumux/opm-common/src/opm/json/JsonObject.cpp:63:79: error: 'string' is unavailable: introduced in macOS 10.15
            throw std::invalid_argument("Loading json from file: " + jsonFile.string() + " failed.");
                                                                              ^
/usr/local/opt/llvm/bin/../include/c++/v1/filesystem:998:41: note: 'string' has been explicitly marked unavailable here
  _LIBCPP_INLINE_VISIBILITY std::string string() const { return __pn_; }
                                        ^
```

Does that mean that only macOS >= 10.15 is supported or is something wrong with the version checks for filesystem.

--- a/src/opm/parser/eclipse/Parser/createDefaultKeywordList.cpp
+++ b/src/opm/parser/eclipse/Parser/createDefaultKeywordList.cpp
@@ -17,10 +17,7 @@
   along with OPM.  If not, see <http://www.gnu.org/licenses/>.
 */
 
-#if !defined(WIN32)
-    #define _POSIX_C_SOURCE 200112L
-    #include <stdlib.h>
-#endif
+#include <cstdlib>
 
 #include <iostream>
 #include <fstream>
@@ -31,21 +28,6 @@
 
 
 int main(int , char ** argv) {
-    try {
-        /* sometimes the local env's locales are broken on POSIX. Boost <=
-         * 1.56 uses the std::locale("") constructor which respects user
-         * preferred locales, and might crash. If this is the case, and
-         * we're on a non-windows system (assuming POSIX), in case of an
-         * exception, set the environment to "C" and keep going.
-         *
-         * Can be removed once boost < 1.57 is no longer supported
-         */
-        std::locale( "" );
-    } catch( const std::runtime_error& ) {
-        #if !defined(WIN32)
-        setenv( "LC_ALL", "C", 1 );
-        #endif
-    }
     const char * keyword_list_file = argv[1];
     const char * source_file_path = argv[2];
     const char * init_file_name = argv[3];

